Support python 3.7 docstrings

radomirbosak opened the pull request:
This fixes the change in python API introduced by
https://bugs.python.org/issue29463

Fixes #101.

radomirbosak commented on astor/code_gen.py:
```diff
@@ -226,6 +226,32 @@ def body_or_else(self, node):
+
+        self.newline()
+        self.indentation += indent
+        self.write(repr(value))
```
I'm a bit unsure about using `repr(value)` for representing docstrings since this generates (unusual) single-quote strings:
```python
>>> import ast
>>> import astor
>>> srctxt = '''
def f():
    """docstring"""
    return 2
'''

>>> print(astor.to_source(ast.parse(srctxt)))
def f():
    'docstring'
    return 2
```

I thought about using the `astor.string_repr.pretty_string` function to generate triple-quoted docstrings,

https://github.com/berkerpeksag/astor/blob/b47718fa095e456c064d3d222f296fccfe36266b/astor/string_repr.py#L58

but I failed to understand its arguments
```python
def pretty_string(s, embedded, current_line, uni_lit=False,
                  min_trip_str=20, max_line=100):
```

What should I put as `embedded` and `current_line`?

berkerpeksag replied:
The meaning of `embedded` is documented at https://github.com/berkerpeksag/astor/blob/b47718fa095e456c064d3d222f296fccfe36266b/astor/code_gen.py#L537-L542 (perhaps we should move that snippet into a method or something?)

And see https://github.com/berkerpeksag/astor/blob/b47718fa095e456c064d3d222f296fccfe36266b/astor/code_gen.py#L549-L557 for the meaning of `current_line` (I don't really remember this one, though so I might be wrong.)

radomirbosak commented:
@berkerpeksag I addressed the comments. PTAL

berkerpeksag commented on .travis.yml:
```diff
@@ -8,12 +8,10 @@ python:
   - 3.5
   - 3.6
   - pypy
-  - pypy3.3-5.2-alpha1
```
At the time, that was the only PyPy version that supports a relatively new Python 3 features. We can replace it wtih `pypy3.5` now.

radomirbosak replied:
I replaced the `pypy3.3-5.2-alpha1` with `pypy3.5` in the last commit.

berkerpeksag replied:
Could you rebase this branch now?

berkerpeksag commented on astor/code_gen.py:
```diff
@@ -226,6 +226,32 @@ def body_or_else(self, node):
+                ast.Module, and 1 for ast.FunctionDef, ast.AsyncFunctionDef and
+                ast.ClassDef).
+        """
+        if not sys.version_info >= (3, 7):
```
I'm pretty sure that we are going to keep the `docstring` attribute in the future Pythons so can't we just write `if sys.version_info < (3, 7):` here?

radomirbosak replied:
Ok, I simplified it to
```python
        if sys.version_info < (3, 7):
            return
```

berkerpeksag commented on astor/code_gen.py:
```diff
@@ -226,6 +226,32 @@ def body_or_else(self, node):
 
+    def docstring(self, node, indent=1):
+        """
+        Write the function/module/class docstring attribute
```
Style nit: Please finish the sentence with a period.

radomirbosak replied:
Period added.

berkerpeksag commented on astor/code_gen.py:
```diff
@@ -226,6 +226,32 @@ def body_or_else(self, node):
+        """
+        Write the function/module/class docstring attribute
+
+        This method has an effect only for python >= 3.7 . For python earlier
```
Style nit: 3.7 . -> 3.7. (no space between 3.7 and .)

Style nit: python -> Python

radomirbosak replied:
Space removed.

berkerpeksag commented on astor/code_gen.py:
```diff
@@ -226,6 +226,32 @@ def body_or_else(self, node):
+
+        self.newline()
+        self.indentation += indent
+        self.write('"""%s"""' % _prep_triple_quotes(value))
```
This is just an idea, what do you think about making the following pattern a public function?

```py
'"""%s"""' % _prep_triple_quotes(value)
```

This could be useful for third-party software and we could even reuse it in `pretty_string()`:

https://github.com/berkerpeksag/astor/blob/3c4efdb936cb98892d1915d8d4463b5ba772b935/astor/string_repr.py#L94

radomirbosak replied:
That sounds useful. What should be the function's name?

radomirbosak replied:
By public - did you mean to also export it in `__init__.py`?

berkerpeksag replied:
> By public - did you mean to also export it in __init__.py?

Yes and document it in docs/index.rst. By the way, we can deal with this function in another PR and keep this PR focused on adding Python 3.7 support.

I'll have a look at the rest later today, thanks!

radomirbosak commented:
Rebased to master after merging #103 . Added attributions.

berkerpeksag commented:
I'm planning to release 0.7.0 next week. Do you have time to add some tests? I can take over and finish this PR if you don't have time.

radomirbosak commented:
I added a test for the docstring fix (which fails on current master, but passes on this proposed branch). Are any more tests needed?

And sorry that addressing your comments took so long. I didn't have much time last week.

radomirbosak commented:
Note the changes from today are in form of fixup commits, which need to be squashed before merging.

radomirbosak commented:
I added the tests for class and method docstrings and then one for module docstrings too.

berkerpeksag commented:
Merged in https://github.com/berkerpeksag/astor/commit/524d525e999f98198787bb35d0b18d83cfa25d4a and https://github.com/berkerpeksag/astor/commit/f87ae3b55f9ea403b4c04fb6cd9b0ab675aaa496. Thanks!
